util/collect_metrics_files.rb

```diff
@@ -0,0 +1,253 @@
+    options[:output_dir] = output_dir
+  end
+  opts.on("-t", "--tar_file_name filename", String,
+          "The name of the output tar file ") do |tar_file_name|
+    options[:tar_file_name] = tar_file_name
+  end
+  opts.on("-p", "--poll_interval seconds", String,
+          "metrics_collector module polling interval") do |interval|
+    options[:poll_interval] = interval
+  end
+  opts.on("-a", "--archive_interval seconds", String,
+          "metrics_collector module archiving interval") do |interval|
+    options[:archive_interval] = interval
+  end
+  opts.on("-v", "--verbose", String, "Enable Verbose output") { options[:verbose] = true }
+end.parse!
+
+START_TIME = options[:start_time] || START_TIME_DEFAULT
+END_TIME = options[:end_time] || END_TIME_DEFAULT
+METRICS_DIRECTORY = options[:metrics_dir] || METRICS_DIRECTORY_DEFAULT
+OUTPUT_DIRECTORY = options[:output_dir] || OUTPUT_DIR_DEFAULT
+POLL_INTERVAL = options[:poll_interval] || POLL_INTERVAL_DEFAULT
+ARCHIVE_INTERVAL = options[:archive_interval] || ARCHIVE_INTERVAL_DEFAULT
+TAR_FILE_NAME = options[:tar_file_name] || build_default_tar_filename(START_TIME, END_TIME)
+VERBOSE = options[:verbose] || false
+
+if VERBOSE
+  OPTION_SETTINGS = <<~SETTINGS
+    The following are the resulting options settings:
+      * start: #{START_TIME}
+      * end: #{END_TIME}
+      * metrics_dir: #{METRICS_DIRECTORY}
+      * output_dir: #{OUTPUT_DIRECTORY}
+      * tar_file_name: #{TAR_FILE_NAME}
+      * poll_interval: #{POLL_INTERVAL}
+      * archive_interval: #{ARCHIVE_INTERVAL}
+      * verbose: #{VERBOSE}
+  SETTINGS
+  puts OPTION_SETTINGS
+end
+
+# Process the metrics dir, blindly assume any non-exempted dir is a service dir
+#
+# @author Randell Pelak
+#
+# @param [string] metrics_dir Top level metrics dir
+#
+# @return [void]
+#
+# @example
+#   process_metrics_dir(metrics_dir)
+def process_metrics_dir(metrics_dir)
+  metrics_dir_obj = Dir.open metrics_dir
+  metrics_dir_obj.each do |path|
+    next if path =~ /^scripts$|^bin$|^\.$|^\.\.$/
+
+    absolute_path = "#{metrics_dir_obj.path}/#{path}"
+    service = path
+    next unless File.directory? absolute_path
+
+    process_service_dir(absolute_path, service)
+  end
+end
+
+# Process potential service dirs or subdirs
+#
+# Uses recursion but doesn't care which subdir the files come from, they all go in the
+# same output service dir
+#
+# @author Randell Pelak
+#
+# @param [string] dir_path Service dir or subdir
+# @param [string] service name of the service the directory belongs to
+#
+# @return [void]
+#
+# @example
+#   process_service_dir(dir_path, service)
+# TODO: Would be nice to only make the dir if relevant files were found.  And only do it once.
+def process_service_dir(dir_path, service)
+  service_staging_dir = "#{OUTPUT_DIRECTORY}/#{service}"
+  FileUtils.mkdir(service_staging_dir) unless File.directory?(service_staging_dir)
+  service_dir = Dir.open dir_path
```

Comment:
Sorry, I could not help myself from exploring an alternative implementation using minitar to search for and extract only the json files each tarball contains.

Here is an outline that gathers the json files from the found tarballs, extracts them into a temp directory and calls `process_json_file` on them.  This is not meant to advocate a re-implementation.  I am simply adding it here to consider a couple of things to possibly refactor.

1.  Let `Dir.glob` do the work of walking the file tree.
1.  Use minitar to only get what you need out of the tarball.
 
```
require 'zlib'
require 'minitar'

json_files = Dir.glob("#{service_dir}/**/*.json")
tar_files = Dir.glob("#{service_dir}/**/*.tar.gz")

json_files.each do |jsonfile|
  process_json_file(jsonfile, service)  
end

tar_files.each do |tarfile|
  # use minitar to get only the json files out of the archive
  zr = Zlib::GzipReader.new(File.open(tarfile, 'rb'))
  mtr = Archive::Tar::Minitar::Reader.new(zr)
  arch_jsons = mtr.collect { |x| x.full_name if x.file? && x.full_name.match(/\.json$/) }
  tmp_dir = Dir.mktmpdir
  Archive::Tar::Minitar.unpack(zr, tmp_dir, arch_jsons)
  arch_jsons.each do |a_json|
    process_json_file(File.join(tmp_dir, jsonfile), service)
  end
end
```